Bottom: 37b2a486df3ff6f4466004648f0d169b173d0b2f
Top:    a540c52a792312c36c9cacfbfb23e1cf5c477e0e
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2011-07-25 23:27:29 +0100

hid/gtk: Do not run a main loop from within the netlist window code.

It turns out that this is a really bad idea.. if the main loop is kept
busy, the new main-loop never exits and the events can become reentrant.

I discovered this when testing code for an animation which used the main
event loop for its timing. Each time a node was selected in the netlist
window, stack-frame would grow - as the new nested main loops never
became idle. Segfaults ensued quite readily, presumably due to the code
not being designed to be reentrant.

It appears from reading the code, that these:

  while (gtk_events_pending ())        /* Make sure everything gets built */
    gtk_main_iteration ();

Are not actually necessary.


---

diff --git a/src/hid/gtk/gui-netlist-window.c b/src/hid/gtk/gui-netlist-window.c
index 0f65d8b..31caa2e 100644
--- a/src/hid/gtk/gui-netlist-window.c
+++ b/src/hid/gtk/gui-netlist-window.c
@@ -622,9 +622,6 @@ node_get_node_from_name (gchar * node_name, LibraryMenuType ** node_net)
    */
   ghid_netlist_window_create (gport);
 
-  while (gtk_events_pending ())	/* Make sure everything gets built */
-    gtk_main_iteration ();
-
   /* Now walk through node entries of each net in the net model looking for
      |  the node_name.
    */
@@ -883,9 +880,6 @@ ghid_get_net_from_node_name (gchar * node_name, gboolean enabled_only)
    */
   ghid_netlist_window_create (gport);
 
-  while (gtk_events_pending ())	/* Make sure everything gets built */
-    gtk_main_iteration ();
-
   /* If no netlist is loaded the window doesn't appear. */
   if (netlist_window == NULL)
     return NULL;
